### PR TITLE
[feat] revocate any user/activation issued before a specific date

### PR DESCRIPTION
**Pull request by aricart**
- updated revocation commands to allow for "*" as a possible value to account/user key
- enhanced interactive experience where partial was provided
- added many tests

# Add Revocations

```sh
❯ nsc add account A
[ OK ] generated and stored account key "ACH2KUUI5TGTVNFUW6EYTB7JXSD55ERQASZPYS5KMONPGADAOBX67YXH"
[ OK ] added account "A"
❯ nsc revocations list-users
Error: account A does not have revoked users
```

Revoke all users (no account users)
```sh
❯ nsc revocations add-user -i
? user-public-key public key or '*' to match any user-public-key *
? revoke all credentials created before (0 is now) 0
[ OK ] revoked all users issued before 2020-10-29 15:28:55 -0500 CDT
```
Revoke single user (no account users)
```sh
❯ nsc revocations add-user -i
? user-public-key public key or '*' to match any user-public-key UAUGJSHSTZY4ESHTL32CYYQNGT6MHXDQY6APMFMVRXWZN76RHE2IRN5O
? revoke all credentials created before (0 is now) 0
[ OK ] revoked user "UAUGJSHSTZY4ESHTL32CYYQNGT6MHXDQY6APMFMVRXWZN76RHE2IRN5O"
```

Revoke user by name (no users)
```sh
nsc revocations add-user -n U
Error: user "U" not found
```

Listing users
```
❯ nsc revocations list-users
+------------------------------------------------------------------------------------------+
|                                   Revoked Users for A                                    |
+----------------------------------------------------------+-------------------------------+
| Public Key                                               | Revoke Credentials Before     |
+----------------------------------------------------------+-------------------------------+
| * [All Users]                                            | Thu, 29 Oct 2020 15:28:55 CDT |
| UAUGJSHSTZY4ESHTL32CYYQNGT6MHXDQY6APMFMVRXWZN76RHE2IRN5O | Thu, 29 Oct 2020 15:29:13 CDT |
+----------------------------------------------------------+-------------------------------+
```

Add user and revoke it by name
```sh
❯ nsc add user U
[ OK ] generated and stored user key "UDUV2TECOGWXXNPBTVNPP7QBC6HFRWI7HGLVG3GGW6IBPQYYSTBC6AEC"
[ OK ] generated user creds file `~/.nkeys/creds/O/A/U.creds`
[ OK ] added user "U" to account "A"
❯ nsc revocations add-user -n U
[ OK ] revoked user "UDUV2TECOGWXXNPBTVNPP7QBC6HFRWI7HGLVG3GGW6IBPQYYSTBC6AEC"

```

Reset env, added operator O, and account A - Add an user and revoke it:
```sh
❯ nsc add user U
[ OK ] generated and stored user key "UAJMK6VS5TQDDLBOMQUTEN4YWUYUYY72CHDBHZZPA2ND3ZAWIAY5Y5OA"
[ OK ] generated user creds file `~/.nkeys/creds/O/A/U.creds`
[ OK ] added user "U" to account "A"
❯ nsc revocations add-user -i
? Revoke a user by name Yes
? Select revoked user to clear U
? revoke all credentials created before (0 is now) 0
[ OK ] revoked user "UAJMK6VS5TQDDLBOMQUTEN4YWUYUYY72CHDBHZZPA2ND3ZAWIAY5Y5OA"
```

Revoke all users:
```sh
❯ nsc revocations add-user -u "*" --at 1575158400
[ OK ] revoked all users issued before 2019-11-30 18:00:00 -0600 CST
```

# Delete Users

```sh
❯ nsc revocations delete-user -i
? Select revoked user to clear UAUGJSHSTZY4ESHTL32CYYQNGT6MHXDQY6APMFMVRXWZN76RHE2IRN5O: [Unknown User]
[ OK ] deleted user revocation for "UAUGJSHSTZY4ESHTL32CYYQNGT6MHXDQY6APMFMVRXWZN76RHE2IRN5O"

❯ nsc revocations delete-user -u "*"
[ OK ] deleted all user revocation

❯ nsc revocations delete-user -u "*"
Error: no user revocations set in account A

❯ nsc revocations delete-user -i
Error: no user revocations set in account A
```

# Activations

```sh
❯ nsc add account A
[ OK ] generated and stored account key "AD3YSINK6C6XALVXRZXJYRYLMTNCSSVTGMMQ2BSCTKHKUSZJR5OOZ5CZ"
[ OK ] added account "A"

❯ nsc add export -i
? export type stream
? subject hello.>
? name hello.>
? private stream Yes
[ OK ] added private stream export "hello.>"

❯ nsc add export -i
? export type service
? subject q
? name q
? private service Yes
? track service latency No
? service response type Singleton
[ OK ] added private service export "q"

❯ nsc revocations list-activations
Error: stream hello.> has no revocations

❯ nsc revocations list-activations --service
Error: service q has no revocations
```

Add some revocations:
```sh
❯ nsc revocations add-activation -t "*"
[ OK ] revoked all activations for "hello.>" issued before 2020-10-29 16:09:30 -0500 CDT

nsc revocations add-activation -t ADR7QI6TAFKXLYKJ4V4PVT7DRSXUZQ7KQAYQTP64UT7TOAIM4NENHZNK -s "hello.>"
[ OK ] revoked activation "hello.>" for account ADR7QI6TAFKXLYKJ4V4PVT7DRSXUZQ7KQAYQTP64UT7TOAIM4NENHZNK

❯ nsc revocations add-activation -t "*" --service
[ OK ] revoked all activations for "q" issued before 2020-10-29 16:09:36 -0500 CDT


❯ nsc revocations list-activations
+------------------------------------------------------------------------------------------+
|                           Revoked Accounts for stream hello.>                            |
+----------------------------------------------------------+-------------------------------+
| Public Key                                               | Revoke Credentials Before     |
+----------------------------------------------------------+-------------------------------+
| ADR7QI6TAFKXLYKJ4V4PVT7DRSXUZQ7KQAYQTP64UT7TOAIM4NENHZNK | Thu, 29 Oct 2020 16:29:44 CDT |
| * [All Accounts]                                         | Thu, 29 Oct 2020 16:09:30 CDT |
+----------------------------------------------------------+-------------------------------+


❯ nsc revocations list-activations --service
+--------------------------------------------------+
|          Revoked Accounts for service q          |
+------------------+-------------------------------+
| Public Key       | Revoke Credentials Before     |
+------------------+-------------------------------+
| * [All Accounts] | Thu, 29 Oct 2020 16:09:36 CDT |
+------------------+-------------------------------+

❯ nsc revocations delete-activation -t "*" -s "foo.>"
Error: unable to locate export

❯ nsc revocations delete-activation -t "*" -s "hello.>"
[ OK ] cleared export revocation hello.> for account *

❯ nsc revocations delete-activation -i
? select account A
? is service No
? select stream export hello.>
? select account revocation to delete ADR7QI6TAFKXLYKJ4V4PVT7DRSXUZQ7KQAYQTP64UT7TOAIM4NENHZNK: B
[ OK ] cleared export revocation hello.> for account ADR7QI6TAFKXLYKJ4V4PVT7DRSXUZQ7KQAYQTP64UT7TOAIM4NENHZNK

❯ nsc revocations list-activations
Error: stream hello.> has no revocations

❯ nsc revocations delete-activation -i
? select account A
? is service Yes
? select service export q
? select account revocation to delete *: * [All Accounts]
[ OK ] deleted service q revocation for all accounts
```

### CODE REVIEW

**Comment by aricart** [comment]
Fixes #315

File: cmd/revoke_listactivation_test.go [hunk 1]
```diff
@@ -64,5 +64,112 @@ func TestRevokeListActivation(t *testing.T) {
...
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "doesn't have stream exports")
+}
+
+func TestRevokeListActivationDefaultExport(t *testing.T) {
+	ts := NewTestStore(t, "O")
+	defer ts.Done(t)
+	ts.AddAccount(t, "A")
+	ts.AddExport(t, "A", jwt.Service, "s", false)
+	_, _, err := ExecuteCmd(createRevokeActivationCmd(), "--service", "--target-account", "*")
```
Comment (matthiashanel) [inline cmd/revoke_listactivation_test.go:119]: shouldn't we also look at the claim and find, * in the revocations?
Also, the same for tests below

File: cmd/revoke_listusers_test.go [hunk 2]
```diff
@@ -51,3 +51,31 @@ func TestRevokeListUsers(t *testing.T) {
...
+
+func TestRevokeListUsersAllUsers(t *testing.T) {
+	ts := NewTestStore(t, "revoke_clear_user")
+	defer ts.Done(t)
+	ts.AddAccount(t, "A")
+	_, _, err := ExecuteCmd(createRevokeUserCmd(), "-u", "*")
+	require.NoError(t, err)
+	stdout, _, err := ExecuteCmd(createRevokeListUsersCmd())
+	require.NoError(t, err)
+	require.Contains(t, stdout, "* [All Users]")
```
Comment (matthiashanel) [inline cmd/revoke_listusers_test.go:80]: inspect the claim?

File: cmd/revokeactivation_test.go [hunk 3]
```diff
@@ -157,3 +157,89 @@ func TestRevokeActivationForServiceInteractive(t *testing.T) {
...
+}
+
+func TestRevokeActivationDefaultSubject(t *testing.T) {
+	ts := NewTestStore(t, "O")
+	defer ts.Done(t)
+	ts.AddAccount(t, "A")
+	ts.AddExport(t, "A", jwt.Service, "q", false)
+
+	_, _, err := ExecuteCmd(createRevokeActivationCmd(), "-t", "*", "--service")
+	require.NoError(t, err)
```
Comment (matthiashanel) [inline cmd/revokeactivation_test.go:222]: claim?

File: cmd/revokeclearuser_test.go [hunk 4]
```diff
@@ -77,13 +79,67 @@ func TestRevokeClearUserInteractive(t *testing.T) {
...
+}
+
+func TestClearRevokeAllUsers(t *testing.T) {
+	ts := NewTestStore(t, "O")
+	defer ts.Done(t)
+	ts.AddAccount(t, "A")
+	_, _, err := ExecuteCmd(createRevokeUserCmd(), "-u", "*")
+	require.NoError(t, err)
+	_, _, err = ExecuteCmd(createClearRevokeUserCmd(), "-u", "*")
+	require.NoError(t, err)
```
Comment (matthiashanel) [inline cmd/revokeclearuser_test.go:144]: check revocation list is empty in claim?

File: cmd/revokeuser.go [hunk 5]
```diff
@@ -112,77 +176,38 @@ func (p *RevokeUserParams) PreInteractive(ctx ActionCtx) error {
...
 	r := store.NewDetailedReport(true)
 	StoreAccountAndUpdateStatus(ctx, token, r)
 	if r.HasNoErrors() {
-		r.AddOK("revoked user %s", p.userPubKey)
+		if p.userKey.publicKey == "*" {
+			when := int64(p.at)
+			if when == 0 {
+				when = time.Now().Unix()
+			}
+			r.AddOK("revoked all users issued before %s", time.Unix(when, 0).String())
```
Comment (matthiashanel) [inline cmd/revokeuser.go:207]: wouldn't we expect some cleanup as well?
if I already have revocations and add * makes all revocations for a user prior to when obsolete

File: cmd/revokeuser_test.go [hunk 6]
```diff
@@ -195,3 +195,78 @@ func TestRevokeUserByNkey(t *testing.T) {
...
+	ts := NewTestStore(t, "O")
+	defer ts.Done(t)
+	ts.AddAccount(t, "A")
+	ts.AddUser(t, "A", "U")
+	_, _, err := ExecuteCmd(createRevokeUserCmd(), "--name", "a")
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "not found")
+
+	_, _, err = ExecuteCmd(createRevokeUserCmd(), "--name", "U")
+	require.NoError(t, err)
```
Comment (matthiashanel) [inline cmd/revokeuser_test.go:218]: inspect claim? also applies to tests below

File: cmd/revokeactivation.go [hunk 7]
```diff
@@ -227,7 +220,15 @@ func (p *RevokeActivationParams) Run(ctx ActionCtx) (store.Status, error) {
 	r := store.NewDetailedReport(true)
 	StoreAccountAndUpdateStatus(ctx, token, r)
 	if r.HasNoErrors() {
-		r.AddOK("revoked activation %s for account %s", p.export.Name, p.accountKey.publicKey)
+		if p.accountKey.publicKey == "*" {
```
Comment (matthiashanel) [inline cmd/revokeactivation.go:223]: jwt.All?

Also, when adding * shouldn't you be able to clear all revocations prior to the provided time

File: cmd/revokeuser.go [hunk 8]
```diff
@@ -112,77 +176,38 @@ func (p *RevokeUserParams) PreInteractive(ctx ActionCtx) error {
...
 	token, err := p.claim.Encode(p.signerKP)
 	if err != nil {
 		return nil, err
 	}
-
 	r := store.NewDetailedReport(true)
 	StoreAccountAndUpdateStatus(ctx, token, r)
 	if r.HasNoErrors() {
-		r.AddOK("revoked user %s", p.userPubKey)
+		if p.userKey.publicKey == "*" {
```
Comment (matthiashanel) [inline cmd/revokeuser.go:202]: jwt.All?

**Review by matthiashanel — COMMENTED** [review]
check the actual claims in the unit tests?
When revoking * for activation and user, shouldn't we perform cleanup of older now redundant keys?

File: cmd/revoke_listactivation_test.go (hunk 1 above)
Comment (aricart) [reply cmd/revoke_listactivation_test.go:119]: > check the actual claims in the unit tests?
> When revoking * for activation and user, shouldn't we perform cleanup of older now redundant keys?

That should be done in the JWT library.

File: cmd/revokeactivation.go (hunk 7 above)
Comment (aricart) [reply cmd/revokeactivation.go]: jwt lib should do the cleanup.

File: cmd/revoke_listactivation_test.go (hunk 1 above)
Comment (aricart) [reply cmd/revoke_listactivation_test.go:119]: https://github.com/nats-io/jwt/issues/126

**Review by matthiashanel — APPROVED** [review]
No cleanup of existing revocations when adding *?
- LGTM

**Pull request merged**